node/sync/src/block_sync.rs

```diff
@@ -325,8 +325,21 @@ impl<N: Network> BlockSync<N> {
     /// Returns the next block to process, if one is ready.
     #[inline]
     pub fn process_next_block(&self, next_height: u32) -> Option<Block<N>> {
-        // Try to advance the ledger with a block from the sync pool.
-        self.remove_block_response(next_height)
+        // Acquire the requests write lock.
+        // Note: This lock must be held across the entire scope, due to asynchronous block responses
+        // from multiple peers that may be received concurrently.
+        let requests = self.requests.write();
```

Comment:
```suggestion
        let requests = self.requests.read();
```
reasoning: we don't need to alter the collection in this scope, and I don't see a reason to disallow concurrent readers if by the end of it we're going to produce the same result anyway